core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1698,150 +1269,348 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, time: Time, met
 
-/**
- * Called when replica leader initiates isr change
- */
-class IsrChangeNotificationListener(controller: KafkaController, eventManager: ControllerEventManager) extends IZkChildListener with Logging {
-  override def handleChildChange(parentPath: String, currentChilds: java.util.List[String]): Unit = {
-    import JavaConverters._
-    eventManager.put(controller.IsrChangeNotification(currentChilds.asScala))
-  }
+class PartitionReassignmentHandler(controller: KafkaController, eventManager: ControllerEventManager) extends ZNodeChangeHandler {
+  override val path: String = ReassignPartitionsZNode.path
+
+  override def handleCreation: Unit = eventManager.put(controller.PartitionReassignment)
+
+  override def handleDeletion: Unit = ???
+
+  override def handleDataChange: Unit = eventManager.put(controller.PartitionReassignment)
 }
 
-object IsrChangeNotificationListener {
-  val version: Long = 1L
+class PartitionReassignmentIsrChangeHandler(controller: KafkaController, eventManager: ControllerEventManager, partition: TopicAndPartition) extends ZNodeChangeHandler {
+  override val path: String = TopicPartitionStateZNode.path(partition)
+
+  override def handleCreation: Unit = ???
+
+  override def handleDeletion: Unit = ???
+
+  override def handleDataChange: Unit = eventManager.put(controller.PartitionReassignmentIsrChange(partition))
 }
 
-/**
- * Starts the preferred replica leader election for the list of partitions specified under
- * /admin/preferred_replica_election -
- */
-class PreferredReplicaElectionListener(controller: KafkaController, eventManager: ControllerEventManager) extends IZkDataListener with Logging {
-  override def handleDataChange(dataPath: String, data: Any): Unit = {
-    val partitions = PreferredReplicaLeaderElectionCommand.parsePreferredReplicaElectionData(data.toString)
-    eventManager.put(controller.PreferredReplicaLeaderElection(partitions))
-  }
+class IsrChangeNotificationHandler(controller: KafkaController, eventManager: ControllerEventManager) extends ZNodeChildChangeHandler {
+  override val path: String = IsrChangeNotificationZNode.path
 
-  override def handleDataDeleted(dataPath: String): Unit = {}
+  override def handleChildChange: Unit = eventManager.put(controller.IsrChangeNotification)
 }
 
-class ControllerChangeListener(controller: KafkaController, eventManager: ControllerEventManager) extends IZkDataListener {
-  override def handleDataChange(dataPath: String, data: Any): Unit = {
-    eventManager.put(controller.ControllerChange(KafkaController.parseControllerId(data.toString)))
-  }
+class PreferredReplicaElectionHandler(controller: KafkaController, eventManager: ControllerEventManager) extends ZNodeChangeHandler {
+  override val path: String = PreferredReplicaElectionZNode.path
 
-  override def handleDataDeleted(dataPath: String): Unit = {
-    eventManager.put(controller.Reelect)
-  }
+  override def handleCreation: Unit = eventManager.put(controller.PreferredReplicaLeaderElection)
+
```

Comment:
This will be called when the preferred leader election completes. So, we probably want to make this a no-op instead of throwing an exception.

Reply:
That's a good point. See my above comment describing several ways we can address this.